lib/Transform/Clang/RemoveFirstPrivate.cpp

```diff
@@ -0,0 +1,349 @@
+#include "tsar/Analysis/Memory/Utils.h"
+#include "tsar/Support/MetadataUtils.h"
+#include <clang/Basic/SourceLocation.h>
+#include <llvm/ADT/SmallVector.h>
+#include <llvm/IR/Function.h>
+#include <llvm/IR/Module.h>
+#include <llvm/Support/Debug.h>
+#include <llvm/Support/raw_ostream.h>
+#include "tsar/Transform/Clang/Passes.h"
+#include <bcl/utility.h>
+#include <llvm/Pass.h>
+#include <stack>
+
+using namespace clang;
+using namespace llvm;
+using namespace tsar;
+
+#undef DEBUG_TYPE
+#define DEBUG_TYPE "clang-rfp"
+
+static int getDimensionsNum(QualType qt, std::vector<int>& default_dimensions) {
+  int res = 0;
+  bool sizeIsKnown = true;
+  while(1) {
+    if (qt -> isArrayType()) {
+      auto at = qt -> getAsArrayTypeUnsafe();
+      auto t =  dyn_cast_or_null<ConstantArrayType>(at);
+      if (sizeIsKnown && t) { // get size
+        uint64_t dim = t -> getSize().getLimitedValue();
+        default_dimensions.push_back(dim);
+      }
+      qt = at -> getElementType();
+      res++;
+    } else if (qt -> isPointerType()) {
+        sizeIsKnown = false;
+        qt = qt -> getPointeeType();
+        res++;
+    } else {
+        return res;
+    }
+  }
+}
+
+namespace {
+
+class ClangRemoveFirstPrivate : public FunctionPass, private bcl::Uncopyable {
+public:
+  static char ID;
+
+  ClangRemoveFirstPrivate() : FunctionPass(ID) {
+    initializeClangRemoveFirstPrivatePass(*PassRegistry::getPassRegistry());
+  }
+
+  bool runOnFunction(Function &F) override;
+  void getAnalysisUsage(AnalysisUsage &AU) const override;
+};
+
+struct vars {                   // contains information about variables in
+  bool rvalIsArray = false;     // removefirstprivate clause
+  std::string lvalName;
+  std::string rvalName;
+  int dimensionsNum;
+  std::vector<int> dimensions;
+  std::vector<int> default_dimensions;
+};
+}
+
+char ClangRemoveFirstPrivate::ID = 0;
+
+INITIALIZE_PASS_IN_GROUP_BEGIN(ClangRemoveFirstPrivate, "remove-firstprivate",
+  "Initialize variables in for", false, false,
+  TransformationQueryManager::getPassRegistry())
+INITIALIZE_PASS_DEPENDENCY(TransformationEnginePass)
+INITIALIZE_PASS_DEPENDENCY(ClangGlobalInfoPass)
+INITIALIZE_PASS_IN_GROUP_END(ClangRemoveFirstPrivate, "remove-firstprivate",
+  "Initialize variables in for", false, false,
+  TransformationQueryManager::getPassRegistry())
+
+namespace {
+
+class DeclVisitor : public RecursiveASTVisitor<DeclVisitor> {
+
+  struct DeclarationInfo {
+    DeclarationInfo(Stmt *S) : Scope(S) {}
+
+    Stmt *Scope;
+  };
+public:
+  explicit DeclVisitor(TransformationContext &TfmCtx, const ASTImportInfo &ImportInfo,
+      ClangGlobalInfoPass::RawInfo &RawInfo) :
+    mTfmCtx(&TfmCtx), mImportInfo(ImportInfo),
+    mRawInfo(&RawInfo), mRewriter(TfmCtx.getRewriter()),
+    mContext(TfmCtx.getContext()), mSrcMgr(mRewriter.getSourceMgr()),
+    mLangOpts(mRewriter.getLangOpts()) {}
+
+  bool TraverseStmt(Stmt *S) {
+    if (!S)
+      return true;
+
+    bool ast = false;
```

Comment:
```suggestion
    bool Ast = false;
```
И далее везде и для всех переменных. Есть строгие правила именования переменных.